domain/secret/state/state.go

```diff
@@ -0,0 +1,448 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	coresecrets "github.com/juju/juju/core/secrets"
+	"github.com/juju/juju/domain"
+	domainsecret "github.com/juju/juju/domain/secret"
+	secreterrors "github.com/juju/juju/domain/secret/errors"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// State represents database interactions dealing with storage pools.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new secretMetadata state
+// based on the input database factory method.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// CreateUserSecret creates a user secret, returning an error satisfying [secreterrors.SecretAlreadyExists]
+// if a user secret with the same label already exists.
+func (st State) CreateUserSecret(ctx context.Context, version int, uri *coresecrets.URI, secret domainsecret.UpsertSecretParams) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	revisionUUID, err := uuid.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	checkExistsSQL := "SELECT &secretOwner.secret_id FROM secret_model_owner WHERE label = $secretOwner.label"
```

Comment:
1. the label is optional, so need to check if it's "";
2. the label is not globally unique;
3. it's really just check label but not check secret existence;

Reply:
Good pickup on empty label. Note that this is just for user secrets, so we are just checking the model_secret_owner table, and for that case, the label is unique for the model.